wandb/sdk/launch/runner/kubernetes.py

```diff
@@ -0,0 +1,222 @@
+        self.batch_api = batch_api
+        self.name = name
+        self.namespace = namespace
+        self.job = self.batch_api.read_namespaced_job(name=self.name, namespace=self.namespace)
+
+    @property
+    def id(self) -> str:
+        return self.name
+
+    def wait(self) -> bool:
+        while True:
+            status = self.get_status()
+            wandb.termlog("Job {} status: {}".format(self.name, status))
+            if status != "running":
+                break
+            time.sleep(5)
+        return status == "finished"     # todo: not sure if this (copied from aws runner) is the right approach? should we return false on failure
+
+    def get_status(self) -> Status:
+        job_response = self.batch_api.read_namespaced_job_status(name=self.name, namespace=self.namespace)
+        status = job_response.status
+        # todo: we only handle the 1 pod case. see https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs for multipod handling
+        if status.succeeded == 1:
+            return Status("finished")
+        elif status.failed == 1:
+            return Status("failed")
+        elif status.active == 1:
+            return Status("running")
+        if status.conditions is not None and status.conditions[0].type == "Suspended":
+            return Status("stopped")
+        return Status("unknown")
+
+    def suspend(self) -> None:
+        self.job.spec.suspend = True
+        self.batch_api.patch_namespaced_job(name=self.name, namespace=self.namespace, body=self.job)
+        timeout = TIMEOUT
+        job_response = self.batch_api.read_namespaced_job_status(name=self.name, namespace=self.namespace)
+        while job_response.status.conditions is None and timeout > 0:
+            time.sleep(1)
+            timeout -= 1
+            job_response = self.batch_api.read_namespaced_job_status(name=self.name, namespace=self.namespace)
+        
+        if timeout == 0 or job_response.status.conditions[0].type != "Suspended":
+            raise LaunchError("Failed to suspend job {}. Check Kubernetes dashboard for more info.".format(self.name))
+
+    def cancel(self) -> None:
+        self.suspend()
+        # todo: is that right? suspension will already delete all active pods but suspended jobs can be resumed/uncancelled - should we delete the job
+        
+
+class KubernetesRunner(AbstractRunner):
+    def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:
+        kubernetes = get_module("kubernetes", "KubernetesRunner requires kubernetes to be installed")
+
+        resource_args = launch_project.resource_args.get("kubernetes")
+        if resource_args is None:
+            raise LaunchError("No Kubernetes resource args specified. Specify args via --resource-args with a JSON file or string under top-level key kubernetes")
+        
+        config_file = resource_args.get("config_file")   # kubeconfig, if None then loads default in ~/.kube
```

Comment:
Not sure this is needed if we're running in the kubernetes cluster already.

Reply:
We should likely be using RBAC and k8s service accounts.  Here's an example of loading the config from within the cluster: https://github.com/kubernetes-client/python/blob/master/examples/in_cluster_config.py

Reply:
doing this now